api/src/main/java/org/apache/iceberg/ManifestFile.java

```diff
@@ -49,7 +49,7 @@ public interface ManifestFile {
           Types.LongType.get(),
           "Lowest sequence number in the manifest");
   Types.NestedField SNAPSHOT_ID =
-      optional(
+      required(
```

Comment:
How do we maintain the compatibility of reading old manifestFiles? 

The manifest reader will use new schema and try to read old file and it may fail if the data was not written? 

Are we sure we were always writing the data even for this optional field?

Reply:
> Are we sure we were always writing the data even for this optional field?

I need to check the code for this. Doing something else in parallel today. Will check it tomorrow.

Reply:
Good question. I haven't seen any cases where there is no data. The thing I want to double-check if that we read a required field, but the schema in the field is optional.

Reply:
```
spark-sql ()> select * from examples.nyc_taxi_yellow limit 22;
23/09/29 15:28:54 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
2	2003-03-27 06:07:28	193	2003-03-27 22:30:46	193	1	0.0	1	1	3.31	2.5	0.01	0.0	0.5	0.3	0.0	0.0	N
2	2009-01-02 01:17:42	193	2009-01-02 18:19:52	193	1	0.0	1	1	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	N
2	2009-01-01 08:08:07	193	2009-01-01 08:08:12	193	1	0.0	1	2	3.3	2.5	0.0	0.0	0.5	0.3	0.0	0.0	N
2	2009-01-01 09:34:29	132	2009-01-01 09:52:13	203	1	4.99	1	2	17.3	16.0	0.0	0.0	0.5	0.3	0.0	0.5	N
2	2009-01-01 09:46:10	132	2009-01-01 10:16:52	39	2	9.36	1	2	29.3	28.5	0.0	0.0	0.5	0.3	0.0	0.0	N
2	2009-01-01 11:26:49	7	2009-01-01 23:37:23	7	1	0.0	1	2	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	N
2	2009-01-02 03:33:44	193	2009-01-02 03:34:54	193	1	0.0	1	2	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	N
2	2020-11-30 22:23:38	138	2020-12-01 21:44:12	116	1	8.64	1	1	37.92	26.5	4.0	6.12	0.5	0.3	0.0	0.5	N
NULL	2020-12-01 09:01:00	42	2020-12-01 09:08:00	41	NULL	1.4	NULL	NULL	10.31	9.51	0.0	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:02:49	230	2020-12-01 09:34:29	61	NULL	9.69	NULL	NULL	37.9	34.35	2.75	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:04:00	215	2020-12-01 09:07:00	215	NULL	0.51	NULL	NULL	59.84	56.29	2.75	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:06:00	192	2020-12-01 09:27:00	70	NULL	3.83	NULL	NULL	25.5	24.7	0.0	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:07:00	136	2020-12-01 09:21:00	243	NULL	3.03	NULL	NULL	22.0	18.45	2.75	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:07:28	137	2020-12-01 09:23:35	69	NULL	8.09	NULL	NULL	35.59	32.04	2.75	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:08:00	242	2020-12-01 09:25:00	18	NULL	4.55	NULL	NULL	23.85	20.3	2.75	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:08:00	255	2020-12-01 09:24:00	146	NULL	4.74	NULL	NULL	19.43	15.05	3.58	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:08:37	61	2020-12-01 09:36:29	137	NULL	9.79	NULL	NULL	30.95	27.65	0.0	0.0	0.5	0.3	2.5	0.0	NULL
NULL	2020-12-01 09:09:00	168	2020-12-01 09:20:00	212	NULL	3.52	NULL	NULL	25.5	21.95	2.75	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:12:10	140	2020-12-01 09:31:20	213	NULL	8.99	NULL	NULL	34.82	19.03	2.75	12.24	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:14:00	239	2020-12-01 09:43:00	49	NULL	9.84	NULL	NULL	36.36	28.85	4.21	0.0	0.5	0.3	2.5	0.0	NULL
NULL	2020-12-01 09:16:00	226	2020-12-01 09:41:00	219	NULL	14.79	NULL	NULL	48.17	44.62	2.75	0.0	0.5	0.3	0.0	0.0	NULL
NULL	2020-12-01 09:24:00	140	2020-12-01 09:44:00	116	NULL	5.51	NULL	NULL	30.08	21.9	4.88	0.0	0.5	0.3	2.5	0.0	NULL
Time taken: 16.631 seconds, Fetched 22 row(s)
```

Reply:
It just works :)

Reply:
cc @nastra what do you think?

Reply:
I think optional is correct here. 
V1 uses this then as optional in [V1Metadata](https://github.com/apache/iceberg/blob/9907a97351138e676d08e02ea53b746f4e331ec6/core/src/main/java/org/apache/iceberg/V1Metadata.java#L39).
V2 makes this then required in [V2Metadata](https://github.com/apache/iceberg/blob/9907a97351138e676d08e02ea53b746f4e331ec6/core/src/main/java/org/apache/iceberg/V2Metadata.java#L42)

Reply:
>  It just works

Was it a v1 table? 

But anyways, as @nastra said, we need to support both v1 and v2. Optional seems to be correct Or we should have different classes for v1 and v2.

Reply:
My point is that the code should follow the spec, and not the other way around. However, if we've been writing nullable, then we can't make it required anymore. We just updated it a week ago, but I think that's incorrect: https://github.com/apache/iceberg/pull/8600

On the current docs it is still both required:

![image](https://github.com/apache/iceberg/assets/1134248/29ef71b0-b51f-41f9-b745-1bc5ba4d506d)

```
bin/spark-sql  \
  --packages org.apache.iceberg:iceberg-spark-runtime-3.4_2.12:1.3.1,software.amazon.awssdk:bundle:2.20.18 \
  --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions

...

spark-sql (spielerij)> create table null_or_not_null AS SELECT 1 as id UNION ALL select 2 UNION ALL select 3;
spark-sql (spielerij)> describe table extended null_or_not_null;
id                  	int                 	                    
                    	                    	                    
# Metadata Columns  	                    	                    
_spec_id            	int                 	                    
_partition          	struct<>            	                    
_file               	string              	                    
_pos                	bigint              	                    
_deleted            	boolean             	                    
                    	                    	                    
# Detailed Table Information	                    	                    
Name                	sandbox.spielerij.null_or_not_null	                    
Type                	MANAGED             	                    
Location            	s3://tabular-wh-us-east-1/6efbcaf4-21ae-499d-b340-3bc1a7003f52/45bc0c6c-b104-4b82-8b18-12d0f7acc287	                    
Provider            	iceberg             	                    
Table Properties    	[current-snapshot-id=1823638994492416744,format=iceberg/parquet,format-version=1,manifest-rewrite.submitted=2023-09-29T16:53:31.667746002Z,optimizer.enabled=true,write.delete.parquet.compression-codec=zstd,write.metadata.compression-codec=gzip,write.object-storage.enabled=true,write.parquet.compression-codec=zstd,write.summary.partition-limit=100]
```


Cleared the caches:

```
rm -rf ~/.m2/repository/org/apache/iceberg
rm -rf /Users/fokkodriesprong/.ivy2/cache
```

And confirmed:


```
		::::::::::::::::::::::::::::::::::::::::::::::

		::          UNRESOLVED DEPENDENCIES         ::

		::::::::::::::::::::::::::::::::::::::::::::::

		:: org.apache.iceberg#iceberg-spark-runtime-3.5_2.12;1.4.0-SNAPSHOT: not found

		::::::::::::::::::::::::::::::::::::::::::::::
```

Ran locally `./gradlew publishToMavenLocal`:

```
bin/spark-sql  \                                                                                       
  --packages org.apache.iceberg:iceberg-spark-runtime-3.5_2.12:1.4.0-SNAPSHOT,software.amazon.awssdk:bundle:2.20.18 \
  --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions \


spark-sql ()> SELECT * FROM spielerij.null_or_not_null;
23/09/29 18:59:01 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
1
2
3
Time taken: 8.449 seconds, Fetched 3 row(s)
```

Reply:
> We just updated it a week ago, but I think that's incorrect: https://github.com/apache/iceberg/pull/8600

I thought it was optional from the beginning and hence asked about compatibility. Didn't know it was recently modified. Compatibility shouldn't matter now. 

> My point is that the code should follow the spec, and not the other way around
Agree.

Reply:
Can we also revert the spec change from https://github.com/apache/iceberg/pull/8600 ?

Reply:
I should have mentioned that, I'll be some more checks and sync up with Eduard next week

Reply:
I talked with Ryan to get historical context, and this should be reverted and created a separate PR https://github.com/apache/iceberg/pull/8726 so we can already revert the spec itself.

As in, reading an optional field using a required read schema; this is allowed. I've added a test as well (but probably this is already tested somewhere).